From d3b523af03de40170b0a99ea165f805eaa7f8a52 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Konrad=20Zapa=C5=82owicz?=
 <konrad.zapalowicz@chainguard.dev>
Date: Fri, 5 Sep 2025 11:16:25 +0200
Subject: [PATCH 1/4] refactor(builder): recognize chainguard's distro
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The flux-operator assigns a distro value to an image based on the pull URL.

Currently, it recognizes a couple of flavours of flux-operator that are served
from either AWS or GitHub. As A consequence it errors for any unknown registry,
including Chaninguard's.

This commit changes that behavior by making the code responsible for that
recognize the "chainguard-private" distro that is related to pull URLs from
within the cgr.dev realm.

Signed-off-by: Konrad Zapałowicz <konrad.zapalowicz@chainguard.dev>
---
 internal/builder/images.go | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/internal/builder/images.go b/internal/builder/images.go
index 333d17f..102a770 100644
--- a/internal/builder/images.go
+++ b/internal/builder/images.go
@@ -35,7 +35,6 @@ func ExtractComponentImagesFromObjects(objects []*unstructured.Unstructured, opt
 					cname := container.(map[string]any)["name"].(string)
 					if cname != "manager" {
 						continue
-
 					}
 
 					img := container.(map[string]any)["image"].(string)
@@ -136,8 +135,20 @@ func ExtractComponentImagesWithDigest(srcDir string, opts Options) (images []Com
 		distro = "enterprise-distroless"
 	case "709825985650.dkr.ecr.us-east-1.amazonaws.com/controlplane/fluxcd":
 		distro = "enterprise-distroless"
+	case "cgr.dev/chainguard-private/flux-operator":
+	case "cgr.dev/chainguard-private/flux-operator-fips":
+		distro = "chainguard"
 	default:
-		return nil, fmt.Errorf("unsupported registry: %s", registry)
+		// The customers who are entitled to an image have that served from their
+		// org. Similairly, a customized image will have a differnet name. However,
+		// the registry top-level domain remains the same. This final check is
+		// going to catch all images coming from Chainguard realm no matter the
+		// actual org or image name.
+		if strings.Contains(registry, "cgr.dev") {
+			distro = "chainguard"
+		} else {
+			return nil, fmt.Errorf("unsupported registry: %s", registry)
+		}
 	}
 
 	imageFile := fmt.Sprintf("%s/%s/%s.yaml", srcDir, opts.Version, distro)
-- 
2.50.1

